evm: add fuzz tests

RahulMaganti47 opened the pull request:
Fixes #174.

djb15 commented on evm/test/RateLimit.t.sol:
```diff
@@ -698,4 +701,321 @@ contract TestRateLimit is Test, IRateLimiterEvents {
+    // this should consume capacity on the outbound side
+    // and backfill the inbound side
+    function testFuzz_CircularFlowBackFilling(uint64 mintAmt, uint64 transferAmt) public {
+        vm.assume(transferAmt > 0 && transferAmt < mintAmt);
```
Rather than using `assume` you should try to use `bound` when working with uints. Rather than discarding test cases it just bounds the inputs

djb15 commented on evm/test/RateLimit.t.sol:
```diff
@@ -698,4 +701,321 @@ contract TestRateLimit is Test, IRateLimiterEvents {
+        token.approve(address(nttManager), type(uint256).max);
+        // transfer tokens from user_A -> user_B via the nttManager
+        nttManager.transfer(
+            transferAmount.untrim(decimals), chainId, toWormholeFormat(user_B), false, new bytes(1)
```
Why did you choose to use the `TrimmedAmount` struct to hold these amounts? Doesn't that mean we'll never have an amount > `type(uint64).max` as a `transfer` input?

RahulMaganti47 replied:
i think this is fine for this test (its not really meant to test the queueing logic). can test that more explicitly in a separates test. the amount transferred should be less than the outbound limit. `setOutboundLimit` trims the `uint256` value, which if greater than u64MAX, reverts on `AmountTooLarge.` using trimmed amounts makes the inputs more explicit. to improve this slightly, we could uint256 values as inputs and then trim (and check the revert case).

djb15 replied:
Yeah so my thinking around this wasn't for testing the queuing logic (I agree we don't need to test that here), but more that the trimming works properly and reverts when we expect.

Rather than branching in this fuzz test we could have another fuzz test that always reverts for an expected bounded set of input values?

RahulMaganti47 replied:
yea makes sense to include some more negative tests - will add that in

djb15 commented on evm/test/TrimmedAmount.t.sol:
```diff
@@ -156,4 +156,112 @@ contract TrimmingTest is Test {
+    ) public {
+        // this is guaranteed by trimming
+        uint256 amt = bound(amount, 1, type(uint64).max);
+        vm.assume(fromDecimals <= 18);
```
Could we fuzz with more decimals than this? Think there are some tokens with 24 decimals for example

RahulMaganti47 replied:
yea i tested with a few more decimals. should we add an error if the decimal amount is too large (say 100?). the division becomes very numerically unstable, creating underflow issues.

djb15 replied:
I can't imagine that would ever be a realistic scenario right as so many other protocols would break if a token appeared with a crazy number of decimals! So I think going up to 50 is more than enough coverage, thanks

djb15 commented on evm/test/TrimmedAmount.t.sol:
```diff
@@ -156,4 +156,112 @@ contract TrimmingTest is Test {
+        a.decimals = uint8(bound(a.decimals, 0, 8));
+
+        b.amount = uint64(bound(b.amount, 0, type(uint64).max));
+        b.decimals = uint8(bound(b.decimals, 0, 8));
```
Why only bounding to 8 decimals here and don't think there's any need to bound the amounts since it's already a uint64?

gator-boi commented on evm/test/RateLimit.t.sol:
```diff
@@ -639,7 +641,189 @@ contract TestRateLimit is Test, IRateLimiterEvents {
+
+        // check error conditions
+        if (transferAmount.amount == 0) {
+            vm.expectRevert();
```
Why don't we use the actual revert signature?

djb15 replied:
Also btw @RahulMaganti47 I prefer using `abiEncodeWithSelector` rather than `abiEncodeWithSignature` as then you don't have to copy the error signature string around. Calling that out as I noticed in the `setIsWormholeEvmChain` PR

RahulMaganti47 replied:
ah nice, yea thats cleaner. 

yea we should use the revert signature (it hasn't a standard that we've enforced). there are a few places, esp. in the integration tests where the revert signatures are omitted. i can address those in a subsequent PR

johnsaigle commented on evm/test/TrimmedAmount.t.sol:
```diff
@@ -218,4 +220,106 @@ contract TrimmingTest is Test {
+        vm.assume(toDecimals <= 50);
+
+        // NOTE: this is guaranteeed by trimming
+        vm.assume(fromDecimals <= 8 && fromDecimals <= toDecimals);
```
Can you explain the guarantee here? I'm not finding the relevant code in the TrimmedAmount.sol file (and I don't see it on the Rust side either).

I think it is probably best to remove this check so that the fuzzer can actually guarantee that checks like these are performed.

johnsaigle replied:
As written I believe the fuzzer misses this code path:

https://github.com/wormhole-foundation/example-native-token-transfers/blob/49d5ab0db6ca162e94e757e3e8aec0747ad78b9c/evm/src/libraries/TrimmedAmount.sol#L140-L142

RahulMaganti47 replied:
i don't believe it does? `toDecimals` and `fromDecimals` are flipped here: https://github.com/wormhole-foundation/example-native-token-transfers/blob/92e9f3bc5e9bc26d9d5ab463aa4777a585426401/evm/test/TrimmedAmount.t.sol#L244

johnsaigle replied:
What's the reason for flipping them? The `trim` function should take `fromDecimals` as the first parameter.

Also if they're flipped, then isn't the code missing the inverse branch?

Generally I'm not clear why the two decimal amounts are being constrained like this

RahulMaganti47 replied:
it still goes through the other code path on untrimming but can simplify this without restricting the input

johnsaigle commented:
I wrote fuzz tests for Trimmed Amount here: https://github.com/wormhole-foundation/example-native-token-transfers/pull/269/files#diff-a7a75ec27a39e2e8f3e35eb187347d1adc9eaece37a5abacd6cb2113551a88f6

It would be great if we could keep enforce the same invariants on both sides

RahulMaganti47 commented:
> I wrote fuzz tests for Trimmed Amount here: https://github.com/wormhole-foundation/example-native-token-transfers/pull/269/files#diff-a7a75ec27a39e2e8f3e35eb187347d1adc9eaece37a5abacd6cb2113551a88f6
> 
> It would be great if we could keep enforce the same invariants on both sides

sure - i've updated the existing invariant tests and added in a few more

gator-boi commented on evm/test/RateLimit.t.sol:
```diff
@@ -703,4 +720,369 @@ contract TestRateLimit is Test, IRateLimiterEvents {
+        uint256 totalAmt = (type(uint64).max) / (10 ** decimals);
+        // avoids the ZeroAmount() error
+        // cannot transfer more than what's available
+        vm.assume(transferAmt > 0 && transferAmt <= totalAmt);
```
Use `bound` here?

gator-boi commented on evm/test/TrimmedAmount.t.sol:
```diff
@@ -218,4 +220,153 @@ contract TrimmingTest is Test {
+        uint256 untrimmedAmt = trimmedAmt.untrim(fromDecimals);
+        uint256 untrimmedAmt2 = trimmedAmt2.untrim(fromDecimals);
+
+        // trimming is the left inverse of trimming
```
bad comment?

RahulMaganti47 replied:
no that's right

mdulin2 commented on evm/test/NttManager.t.sol:
```diff
@@ -72,12 +72,20 @@ contract TestNttManager is Test, IRateLimiterEvents {
-        assertEq(countSetBits(16), 1);
-        assertEq(countSetBits(65535), 16);
+    // naive implementation of countSetBits to test against
+    function simpleCount(uint64 n) public pure returns (uint8) {
```
For what it's worth, I tested the 'countSetBits()' in Python using the implementation and feel pretty good about it as well.

RahulMaganti47 replied:
tested just in case

mdulin2 commented on evm/test/RateLimit.t.sol:
```diff
@@ -703,4 +720,369 @@ contract TestRateLimit is Test, IRateLimiterEvents {
+        require(expectedRevertHash == actualRevertHash, "call did not revert as expected");
+    }
+
+    // transfer tokens from user_A to user_B
```
Good looking test!

mdulin2 commented on evm/test/RateLimit.t.sol:
```diff
@@ -703,4 +720,369 @@ contract TestRateLimit is Test, IRateLimiterEvents {
+
+        // allow for amounts greater than uint64 to check if [`setOutboundLimit`] reverts
+        // on amounts greater than u64 MAX.
+        if (mintAmt.scale(decimals, 8) > type(uint64).max) {
```
Will this kill a bunch of runs of fuzzing?

RahulMaganti47 replied:
this case should revert for sure anyway

RahulMaganti47 replied:
this case should revert in anyway
